src/Turing.jl

```diff
@@ -121,6 +119,15 @@ Sampler(alg) = Sampler(alg, Selector())
 Sampler(alg, model::Model) = Sampler(alg, model, Selector())
 Sampler(alg, model::Model, s::Selector) = Sampler(alg, model, s)
 
+abstract type AbstractContext end
```

Comment:
I feel this context thing is like what @yebai wanted to do via `AbstractRunner`. But the difference here is that for context we "dispatch" during compile time while runner "dispatch" during runtime. 
1. Is my understanding correct here?
2. If so, we probably only want to keep such functionality at one place.

Reply:
Also is this functionality is only possible to be done in compile time?

Reply:
I think it is similar to the `AbstractRunner` concept but I never quite understood the objective of `AbstractRunner`. The point of the context here is that it introduces another dispatch layer that lets us modify the behavior of `~`. But instead of getting involved with the `assume` and `observe` dispatch on the sampler type, I made the function `assume_or_observe` which dispatches on the output of `@preprocess` to call either `assume` or `observe`, and simultaneously dispatches on the context to modify the inputs or outputs to the `assume` or `observe` functions as needed.

I don't think the concepts of compile-time and run-time are relevant here. We are dispatching on the context to run different code for different contexts. This naturally means that different code will be compiled and run for different contexts.

Reply:
I guess we can achieve the same thing by dispatching over a different sampler type directly can we? I would agree having a context is much cleaner though.